Add key password for Android export and configured signature options

lph993 opened the pull request:
For the signature options exported and configured by Android, the keystore key password option is added to solve the problem that the key in a keystore has a password. When the key password is empty, it is compatible with the previous code.

<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

lph993 committed:
Add key password for Android export and configured signature options

For the signature options exported and configured by Android, the keystore key password option is added to solve the problem that the key in a keystore has a password. When the key password is empty, it is compatible with the previous code.

lph993 committed:
Merge branch 'master' into master

timothyqiu commented on platform/android/export/export_plugin.cpp:
```diff
@@ -2383,6 +2392,12 @@ Error EditorExportPlatformAndroid::sign_apk(const Ref<EditorExportPreset> &p_pre
 	args.push_back("--ks-key-alias");
 	args.push_back(user);
+
+	if(!userPassword.empty()) {
```
```suggestion
	if (!userPassword.empty()) {
```

lph993 replied:
already edited.
Issue 1: Since property names are changed / added, error messages below should be changed accordingly. e.g.
Answer: Corresponding processing has been done in the code. When "user_password" is empty, "password" will be used as the default password, which is compatible with the previous processing method of the document author.
Refer to the following code:
```
// Fix export or run program signature issues.
args.push_back("--key-pass");
if (user_password.is_empty()) {
args.push_back("pass:" + password);
} else {
args.push_back("pass:" + user_password);
}

```
Question 2: Why is this called "User Pass" instead of "Key Pass"?
Answer: The documentation author used "user" for "--ks-key-alias", and to correspond with "user_password" for "--key-pass" .
--This text is translated from the Internet.

timothyqiu commented on platform/android/export/export_plugin.cpp:
```diff
@@ -2089,7 +2091,8 @@ bool EditorExportPlatformAndroid::can_export(const Ref<EditorExportPreset> &p_pr
 	String dk_user = p_preset->get("keystore/debug_user");
-	String dk_password = p_preset->get("keystore/debug_password");
+	String dk_password = p_preset->get("keystore/debug_keystore_password");
+	String dk_user_password=p_preset->get("keystore/debug_user_password");
```
```suggestion
	String dk_user_password = p_preset->get("keystore/debug_user_password");
```

timothyqiu reviewed (changes requested):
Please make sure to format your code according to `.clang-format`. The suggestions are two examples.

You can run `clang-format` locally and config it as a commit hook. See https://docs.godotengine.org/zh_CN/stable/community/contributing/code_style_guidelines.html#c-and-objective-c

lph993 committed:
Update export_plugin.cpp

Add the alias password option of Android signature command to solve the problem that the keystore password is different from the alias password.

lph993 commented:
OK



---Original---
From: "Haoyu ***@***.***&gt;
Date: Sat, Apr 9, 2022 19:16 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [godotengine/godot] Add key password for Android export and configured signature options (PR #60055)




 
@timothyqiu requested changes on this pull request.
 
Please make sure to format your code according to .clang-format. The suggestions are two examples.
 
You can run clang-format locally and config it as a commit hook. See https://docs.godotengine.org/zh_CN/stable/community/contributing/code_style_guidelines.html#c-and-objective-c

 
In platform/android/export/export_plugin.cpp:
 &gt; @@ -2383,6 +2392,12 @@ Error EditorExportPlatformAndroid::sign_apk(const Ref<EditorExportPreset&gt; &amp;p_pre  args.push_back("pass:" + password);  args.push_back("--ks-key-alias");  args.push_back(user); + +if(!userPassword.empty()) {  ⬇️ Suggested change -if(!userPassword.empty()) { +if (!userPassword.empty()) {  
 
In platform/android/export/export_plugin.cpp:
 &gt; @@ -2089,7 +2091,8 @@ bool EditorExportPlatformAndroid::can_export(const Ref<EditorExportPreset&gt; &amp;p_pr    String dk = p_preset-&gt;get("keystore/debug");  String dk_user = p_preset-&gt;get("keystore/debug_user"); -String dk_password = p_preset-&gt;get("keystore/debug_password"); +String dk_password = p_preset-&gt;get("keystore/debug_keystore_password"); +String dk_user_password=p_preset-&gt;get("keystore/debug_user_password");  ⬇️ Suggested change -String dk_user_password=p_preset-&gt;get("keystore/debug_user_password"); +String dk_user_password = p_preset-&gt;get("keystore/debug_user_password");  
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

lph993 committed:
Update export_plugin.cpp

Added Android signature command option "--key-pass" to solve the apk signature problem. Compatible with previous signature methods.

lph993 committed:
Update export_plugin.cpp

Added Android signature command option "--key-pass" to solve the apk signature problem. Compatible with previous signature methods.

lph993 committed:
Update export_plugin.cpp

Added Android signature command option "--key-pass" to solve the apk signature problem. Compatible with previous signature methods.

timothyqiu commented on platform/android/export/export_plugin.cpp:
```diff
@@ -2343,15 +2348,18 @@ Error EditorExportPlatformAndroid::sign_apk(const Ref<EditorExportPreset> &p_pre
 	String keystore;
 	String password;
 	String user;
+	String userPassword;
```
```suggestion
	String user_password;
```

Godot uses `snake_case` in C++. The same applies to other variable names below.

lph993 replied:
already edited.

timothyqiu commented on platform/android/java/app/config.gradle:
```diff
@@ -289,6 +289,14 @@ ext.getDebugKeyAlias = { ->
 
+ext.getDebugAliasPassword={->
+    String keyAliasPassword = project.hasProperty("debug_keystore_alias_password") ? project.property("debug_keystore_alias_password") : ""
+    if(keyAliasPassword==null||keyAliasPassword.isEmpty()){
```
I'm not sure if we have style guide for Gradle files, but I think it's better to use the same format as codes around this.

lph993 replied:
already edited.

lph993 replied:
already edited.

timothyqiu commented on platform/android/export/export_plugin.cpp:
```diff
@@ -1704,11 +1704,13 @@ void EditorExportPlatformAndroid::get_export_options(List<ExportOption> *r_optio
+	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "keystore/debug_keystore_password"), ""));
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "keystore/debug_user"), ""));
-	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "keystore/debug_password"), ""));
+	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "keystore/debug_user_password"), ""));
```
Since property names are changed / added, error messages below should be changed accordingly. e.g.

> "Either Release Keystore, Release User AND Release Password settings must be configured OR none of them."

lph993 replied:
Corresponding processing has been done in the code. When "user_password" is empty, "password" will be used as the default password, which is compatible with the previous processing method of the document author.
Refer to the following code:
```
// Fix export or run program signature issues.
args.push_back("--key-pass");
if (user_password.is_empty()) {
args.push_back("pass:" + password);
} else {
args.push_back("pass:" + user_password);
}
```

timothyqiu replied:
I mean, this PR replaced "Debug Password" with "Debug Keystore Password" and "Debug User Password", and "Debug User Password" is optional.

The error message here still uses "Debug Password":

> Either Debug Keystore, Debug User AND Debug Password settings must be configured OR none of them.

So I think you should replace the "Debug Password" here with "Debug Keystore Password" too.

lph993 replied:
ok,already edited.

lph993 replied:
Boss, I just found out that you are from China, should I answer in Chinese? English is not much.

timothyqiu replied:
It's better to use English, so others can understand.

lph993 commented:
The code has been modified.


---Original---
From: "Haoyu ***@***.***&gt;
Date: Sat, Apr 9, 2022 19:16 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [godotengine/godot] Add key password for Android export and configured signature options (PR #60055)




 
@timothyqiu requested changes on this pull request.
 
Please make sure to format your code according to .clang-format. The suggestions are two examples.
 
You can run clang-format locally and config it as a commit hook. See https://docs.godotengine.org/zh_CN/stable/community/contributing/code_style_guidelines.html#c-and-objective-c

 
In platform/android/export/export_plugin.cpp:
 &gt; @@ -2383,6 +2392,12 @@ Error EditorExportPlatformAndroid::sign_apk(const Ref<EditorExportPreset&gt; &amp;p_pre  args.push_back("pass:" + password);  args.push_back("--ks-key-alias");  args.push_back(user); + +if(!userPassword.empty()) {  ⬇️ Suggested change -if(!userPassword.empty()) { +if (!userPassword.empty()) {  
 
In platform/android/export/export_plugin.cpp:
 &gt; @@ -2089,7 +2091,8 @@ bool EditorExportPlatformAndroid::can_export(const Ref<EditorExportPreset&gt; &amp;p_pr    String dk = p_preset-&gt;get("keystore/debug");  String dk_user = p_preset-&gt;get("keystore/debug_user"); -String dk_password = p_preset-&gt;get("keystore/debug_password"); +String dk_password = p_preset-&gt;get("keystore/debug_keystore_password"); +String dk_user_password=p_preset-&gt;get("keystore/debug_user_password");  ⬇️ Suggested change -String dk_user_password=p_preset-&gt;get("keystore/debug_user_password"); +String dk_user_password = p_preset-&gt;get("keystore/debug_user_password");  
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

timothyqiu commented on platform/android/export/export_plugin.cpp:
```diff
@@ -2383,6 +2392,15 @@ Error EditorExportPlatformAndroid::sign_apk(const Ref<EditorExportPreset> &p_pre
 	args.push_back("--ks-key-alias");
 	args.push_back(user);
+	
+	//Fix export or run program signature issues.
+	args.push_back("--key-pass");
+	if (!userPassword.is_empty()) {
+		args.push_back("pass:" + userPassword);
+	} else {
+		args.push_back("pass:" + password);
+	}
```
It would be more readable to swap the branches, avoiding the "not".

Also, please start the comment with `// ` and end it with a period :)

```suggestion
	// Fix export or run program signature issues.
	args.push_back("--key-pass");
	if (userPassword.is_empty()) {
		args.push_back("pass:" + password);
	} else {
		args.push_back("pass:" + userPassword);
	}
```

timothyqiu commented on platform/android/export/export.cpp:
```diff
@@ -47,6 +47,7 @@ void register_android_exporter() {
 	EditorSettings::get_singleton()->add_property_hint(PropertyInfo(Variant::STRING, "export/android/debug_keystore", PROPERTY_HINT_GLOBAL_FILE, "*.keystore,*.jks"));
 	EDITOR_DEF("export/android/debug_keystore_user", "androiddebugkey");
 	EDITOR_DEF("export/android/debug_keystore_pass", "android");
+	EDITOR_DEF("export/android/debug_keystore_user_pass", "android");
```
Why is this called "User Pass" instead of "Key Pass"?

lph993 replied:
The documentation author used "user" for "--ks-key-alias", and to correspond with "user_password" for "--key-pass" .

lph993 replied:
The Android export documentation does need to be changed. I personally feel that developers should not be forced to use the same password to sign the apk. This is wrong. For example, use the newly generated keystore file to sign for updating the same package with different names. I guess the APP signed by the keystore file cannot be updated and replaced, because the signature fingerprints are different, which is not conducive to the migration of Android projects migrated from other IDEs.
Moreover, the command parameter "--key-pass" was not added, which caused a signature error when godot ran to an Android device.
The password of the keystore is used as the alias password in order not to conflict with the author's intention as much as possible, although the author's intention is not understood.
I'm not very good at English, I hope you can edit the docs document.

lph993 replied:
Use the password of the keystore as the default password when the key alias password is empty.
Sorry, I translated my text from the Internet, and there may be parts that are not easy to understand.

lph993 committed:
Update export_plugin.cpp

lph993 committed:
Update config.gradle

lph993 committed:
Update export_plugin.cpp

lph993 committed:
Update export_plugin.cpp

lph993 committed:
Update export_plugin.cpp

lph993 committed:
Merge branch 'godotengine:master' into master

lph993 committed:
Update export_plugin.cpp

timothyqiu commented:
In the current [documentation](https://docs.godotengine.org/en/stable/tutorials/export/exporting_for_android.html#exporting-for-google-play-store):

> **Release User:** Replace with the key alias.
> **Release Password:** Key password. Note that the keystore password and the key password currently have to be the same.

The "Password" here is the *key password*. It makes sense as "User" refers to the key and "Password" refers to its password. It's the keystore password that has to be the same as the key password currently.

So I think it's the "keystore password" that should be optional and fallback to "key password". But since you changed the name of both fields, I doubt if it's worth to be made optional at all :thinking:

lph993 committed:
Update export_plugin.cpp

lph993 committed:
Update .clang-tidy

lph993 committed:
Merge remote-tracking branch 'upstream/master'

lph993 committed:
Merge remote-tracking branch 'upstream/master'

lph993 committed:
Merge branch 'master' of github.com:lph993/godot

lph993 committed:
Already edited.

lph993 committed:
Merge branch 'master' into master

m4gr3d commented on .clang-tidy:
```diff
@@ -1,7 +1,7 @@
 ---
-Checks: 'clang-diagnostic-*,clang-analyzer-*,-*,modernize-redundant-void-arg,modernize-use-bool-literals,modernize-use-default-member-init,modernize-use-nullptr,readability-braces-around-statements,readability-redundant-member-init'
+Checks:          'clang-diagnostic-*,clang-analyzer-*,-*,modernize-redundant-void-arg,modernize-use-bool-literals,modernize-use-default-member-init,modernize-use-nullptr,readability-braces-around-statements'
 WarningsAsErrors: ''
-HeaderFilterRegex: ''
+HeaderFilterRegex: '.*'
```
Are those changes intentional?

m4gr3d reviewed (changes requested):
The rest of the logic looks good, although I'd say since we're not worried about conserving backward compatibility for Godot 4, let's go ahead and rename the variables (and update the documentation) to reflect the proper Android usage, otherwise we may introduce confusion for users that attempt to reference the actual Android documentation.

So the keystore password should clearly be referenced as such, and the key alias and key alias password should clearly be referenced as well and replace the `user` term we've been using.
